### ISSUE TITLE
The profiler crashes the program when trying to stop on different threads without context manager

### ISSUE THREAD

**Issue by gcramer23** [pytorch/pytorch#64795]
## 🐛 Bug

The profiler crashes the program when attempting to stop profilers on different threads.

## To Reproduce

```python

import threading
import time

from torch.profiler import ProfilerActivity, profile


def basic_profiler():
    p = profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA])
    p.start()
    print(f"profiler start thread={threading.get_ident()}")
    time.sleep(2)
    p.stop()
    print(f"profiler stop thread={threading.get_ident()}")


def main():
    threads = []
    for i in range(2):
        time.sleep(2)
        threads.append(threading.Thread(target=basic_profiler))
        threads[i].start()
    for i in range(2):
        threads[i].join()


if __name__ == "__main__":
    main()

```

```
[New Thread 0x7ffefa324700 (LWP 28209)]
profiler start thread=140732719757056
[New Thread 0x7fff02b25700 (LWP 28218)]
profiler start thread=140733238630144
profiler stop thread=140732719757056
[Thread 0x7ffee3c4f700 (LWP 28199) exited]

Thread 19 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff02b25700 (LWP 28218)]
0x00007fffbdd3ad7c in libkineto::ActivityProfiler::processTraceInternal(libkineto::ActivityLogger&) ()
   from /fsx/users/gcramer/work/pytorch/torch/lib/libtorch_cpu.so
```

## Expected behavior

We can stop the profilers.

## Environment

Collecting environment information...
PyTorch version: 1.10.0a0+gite133919
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.21.1

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: 
GPU 0: A100-SXM4-40GB
GPU 1: A100-SXM4-40GB
GPU 2: A100-SXM4-40GB
GPU 3: A100-SXM4-40GB
GPU 4: A100-SXM4-40GB
GPU 5: A100-SXM4-40GB
GPU 6: A100-SXM4-40GB
GPU 7: A100-SXM4-40GB

Nvidia driver version: 450.119.03
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

**Comment by robieta** [comment]
Right now there can only be one active profiler due to an implementation detail. (At the very least, we should enforce that somewhere so the profiler doesn't segfault...) Is there a distributed application for having multiple profilers active at once?

**Comment by guotuofeng** [comment]
From pytorch lightning point of view, it has the possibility that multiple profiler instance active in case of prefetching feature. 

For e.g., one profiler for batch one and another profiler for batch 2.

**Comment by gcramer23** [comment]
>  Is there a distributed application for having multiple profilers active at once?

The RPC framework https://pytorch.org/docs/stable/rpc.html when using the TensorpipeAgent, has a thread pool of worker threads. I want to profile on the worker threads that execute tasks on a process. For example, profile a parameter server that is located on a device. I encountered the crash when trying to stop a profiler that I started a worker thread 2.

A separate issue to profile all worker threads on a process  https://github.com/pytorch/pytorch/issues/64799.

**Comment by pritamdamania87** [comment]
> The RPC framework https://pytorch.org/docs/stable/rpc.html when using the TensorpipeAgent, has a thread pool of worker threads. I want to profile the worker threads that execute tasks on a process. For example, profile a parameter server that is located on a device. I encountered the crash when trying to stop a profiler that I started on worker thread 2.

@rohan-varma Shouldn't our RPC profiling work be reusable here? IIUC, I don't think @gcramer23 needs to manually create profilers per thread?

**Comment by rohan-varma** [comment]
@pritamdamania87 @gcramer23 Yep, as discussed in https://github.com/pytorch/pytorch/issues/64799, this should work for `torch.autograd.profiler` but we haven't built out support for doing this in `torch.profiler` AFAIK. 

> Right now there can only be one active profiler due to an implementation detail. (At the very least, we should enforce that somewhere so the profiler doesn't segfault...) I

@robieta @ilia-cher Can you elaborate on why this is the case? AFAIK the profiler is thread local, so there should ideally be no issues in enabling distinct profilers across threads and individually aggregating their results?
